RFC-0003-arrow-flight-connector.md

```diff
@@ -0,0 +1,83 @@
+# **RFC0003 for Presto**
+
+## Arrow Flight connector template in Presto
+
+Proposers
+
+* Elbin Pallimalil
+
+## [Related Issues]
+
+https://github.com/prestodb/presto/issues/22729
+
+## Summary
+
+Create a base Arrow Flight module in Presto which can be extended to create an Arrow Flight connector.
+
+## Background
+
+Apache Arrow Flight is a new general-purpose client-server framework that simplifies high performance transport of large datasets over network interfaces. https://arrow.apache.org/blog/2019/10/13/introducing-arrow-flight/
+
+Presto currently supports different datasources through connectors for each type of datasource. The connectors can use JDBC or some other protocol for datasource connection. If we use Flight client to connect to the different datasources we can improve the performance of the datasource connection. Also we would only need to develop one Flight client that can be used across different datasources instead of developing individual connectors for each type of datasource.
+
+## Proposed Implementation
+
+A template for Flight connector will be created as `presto-base-arrow-flight` module. Since each implementation of flight server will be different, this base module can be extended to create a flight connector that supports a particular implementation. The template will support only read operations.
+
+### High level design
+
+#### Proposed design
+
+![Design diagram](arrow-flight-connector/Presto-flight-client.drawio.png)
+
+The Arrow Flight libraries provide a development framework for implementing a service that can discover metadata and send and receive data streams. Flight APIs are used to retrieve the data from the selected table. The steps below need to be followed. 
+
+1. Create a Flight Client using the host, port, ssl information of the Flight service where the route is opened. 
+2. Use the Flight Client to authenticate to the Flight service. 
+3. Create a FlightDescriptor using a command
+    1. The connector template will define an abstract class that returns a byte array.
+    1. The concrete implementation of this connector template will define a class that returns a byte array for the `FlightDescriptor.command` method. 
+    2. The concrete implementation will specify the necessary details for the database connection in the `command` byte array.
+4. Use the FlightDescriptor to obtain a FlightInfo object. 
+5. Use the FlightInfo object to obtain the list of endpoints we will be retrieving data from. 
+6. Use each endpoint to obtain a Ticket. 
+7. Use the Ticket to obtain a Stream. 
+8. Use the Stream to read data. 
+    1. Get VectorSchemaRoot from Flight stream
+    2. For each column in the result, build a Block from the FieldVector
+    3. Return a page with all the constructed Blocks.
```

Comment:
So a page will roughly correspond to an Arrow RecordBatch. Is there some configuration planned to set the RecordBatch size or will it be specific for the connector/datasource being used?

Reply:
https://arrow.apache.org/docs/java/reference/org/apache/arrow/flight/FlightStream.html#next-- We are getting the stream using `FlightStream.next()`. I couldn't find an option to specify the size of Vector data returned by the stream. So I assume this Vector data size will be determined by the datasource or the flight service.

Reply:
There isn't a built-in option for this, but it's usually a parameter that could help tune performance on whatever is consuming the Arrow batches. In this case it sounds like it would be useful to tune the conversion from Arrow to a page, depending on the data types and width of the records. So it might be something to keep in mind since it would apply to all datasources.